test: 32-bit Clang `ipc_test` failure at `-O0`

Noticed as part of this branch #29796, however it can also be reproduced with master (8fa10edcd1706a1f0dc9d8c3adbc8efa3c7755bf) by reproducing the equivalent CI & setting C(XX)FLAGS to -O0:
```bash
make -C depends/ MULTIPROCESS=1 NO_QT=1 NO_WALLET=1 NO_ZMQ=1 NO_USDT=1 CFLAGS="-O0" CXXFLAGS="-O0" DEBUG=1 -j19 HOST=i686-pc-linux-gnu
cmake -B build --toolchain /root/ci_scratch/depends/i686-pc-linux-gnu/toolchain.cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER='clang;-m32' -DCMAKE_CXX_COMPILER='clang++;-m32'
cmake --build build -j18
```

```bash
./build/src/test/test_bitcoin --run_test=ipc*
Running 2 test cases...
terminate called after throwing an instance of 'kj::ExceptionImpl'
  what():  /root/ci_scratch/depends/i686-pc-linux-gnu/include/kj/common.h:1797: failed: expected start <= end && end <= size_; Out-of-bounds ArrayPtr::slice().
stack: 5ca0dd6d 5c78b5db 5c8cc599 5ca0adc9 5c94fb37 5c950070 5c9cd709 5c9cf58a 5c9d72fd 5c9d025e 5c776f11 5bf6351d 5bf6347d 5bf633cd 5bf6337b 5bf6331d 5bf63194 f7afa4b0 f773dff6 f77d55b7
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
    ??:0: returning here
unknown location(0): fatal error: in "ipc_tests/ipc_tests": signal: SIGABRT (application abort requested)
test/ipc_tests.cpp(12): last checkpoint: "ipc_tests" test entry
test_bitcoin: common/args.cpp:578: void ArgsManager::AddArg(const std::string &, const std::string &, unsigned int, const OptionsCategory &): Assertion `ret.second' failed.
unknown location(0): fatal error: in "ipc_tests/parse_address_test": signal: SIGABRT (application abort requested)
test/ipc_tests.cpp(20): last checkpoint: "parse_address_test" fixture ctor

*** 2 failures are detected in the test module "Bitcoin Core Test Suite"
```

I thought I tracked down the problem, and some change I made to the test caused it to pass, but rerunning the the fix in a new container, it no longer works, so I have to go back.

Here were notes I was about to post about potential problem & fix. In any case I suspect problem would be avoided if we just consistently used gcc or clang for this build and didn't try to mix them.

---

Test failure seems to be caused by the container using incompatible ABI versions for the depends build and the main build. The depends build is using gcc with ABI version 18:

```bash
gcc -E -x c++ - -dM <<< "" | grep ABI
#define __GXX_ABI_VERSION 1018
```

The bitcoin build is using clang with ABI version 2:

```bash
clang++ -E -x c++ - -dM <<< "" | grep ABI
#define __GXX_ABI_VERSION 1002
```

In clang, the ABI version is determined when the compiler is built, but in gcc you can control it with the `-fabi-version` option, so the following change seems to fix the build:

```diff
--- a/depends/hosts/linux.mk
+++ b/depends/hosts/linux.mk
@@ -1,5 +1,5 @@
 linux_CFLAGS=-pipe -std=$(C_STANDARD)
-linux_CXXFLAGS=-pipe -std=$(CXX_STANDARD)
+linux_CXXFLAGS=-pipe -std=$(CXX_STANDARD) -fabi-version=2
 
 ifneq ($(LTO),)
 linux_AR = $(host_toolchain)gcc-ar
--- a/depends/packages/libmultiprocess.mk
+++ b/depends/packages/libmultiprocess.mk
@@ -27,3 +27,5 @@ endef
 define $(package)_stage_cmds
   $(MAKE) DESTDIR=$($(package)_staging_dir) install-lib
 endef
+
+$(package)_cxxflags += -fabi-version=11

```

First part of the diff setting ` -fabi-version=2` in the depends build is the main fix. The second part of the diff setting ` -fabi-version=11` is a workaround for an issue that happened specifically with the multiprocess package. Because when `-fabi-version` is set to 10 or below there is a compile error:


<details><summary>error: no matching function for call to ‘std::__uniq_ptr_data</summary>
<p>

```c++
[ 20%] Building CXX object CMakeFiles/mputil.dir/src/mp/util.cpp.o
In file included from /usr/include/c++/13/bits/shared_ptr_base.h:59,
                 from /usr/include/c++/13/bits/shared_ptr.h:53,
                 from /usr/include/c++/13/condition_variable:45,
                 from /usr/include/c++/13/future:41,
                 from /ci_container_base/depends/work/build/i686-pc-linux-gnu/libmultiprocess/07c917f7ca910d66abc6d3873162fc9061704074-722fc6d9234/include/mp/util.h:11,
                 from /ci_container_base/depends/work/build/i686-pc-linux-gnu/libmultiprocess/07c917f7ca910d66abc6d3873162fc9061704074-722fc6d9234/src/mp/util.cpp:6:
/usr/include/c++/13/bits/unique_ptr.h: In instantiation of ‘constexpr std::unique_ptr<_Tp, _Dp>::unique_ptr() [with _Del = std::__future_base::_Result_base::_Deleter; <template-parameter-2-2> = void; _Tp = std::__future_base::_Result_base; _Dp = std::__future_base::_Result_base::_Deleter]’:
/usr/include/c++/13/future:340:34:   required from here
/usr/include/c++/13/bits/unique_ptr.h:305:11: error: no matching function for call to ‘std::__uniq_ptr_data<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter, true, true>::__uniq_ptr_data()’
  305 |         : _M_t()
      |           ^~~~~~
/usr/include/c++/13/bits/unique_ptr.h:241:40: note: candidate: ‘template<class _Del> std::__uniq_ptr_data<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter, true, true>::__uniq_ptr_data(std::__uniq_ptr_impl<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>::pointer, _Del&&) [inherited from std::__uniq_ptr_impl<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>]’
  241 |       using __uniq_ptr_impl<_Tp, _Dp>::__uniq_ptr_impl;
      |                                        ^~~~~~~~~~~~~~~
/usr/include/c++/13/bits/unique_ptr.h:241:40: note:   template argument deduction/substitution failed:
/usr/include/c++/13/bits/unique_ptr.h:305:11: note:   candidate expects 2 arguments, 0 provided
  305 |         : _M_t()
      |           ^~~~~~
/usr/include/c++/13/bits/unique_ptr.h:241:40: note: candidate: ‘std::__uniq_ptr_data<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter, true, true>::__uniq_ptr_data(std::__uniq_ptr_impl<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>::pointer) [inherited from std::__uniq_ptr_impl<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter>]’
  241 |       using __uniq_ptr_impl<_Tp, _Dp>::__uniq_ptr_impl;
      |                                        ^~~~~~~~~~~~~~~
/usr/include/c++/13/bits/unique_ptr.h:241:40: note:   candidate expects 1 argument, 0 provided
/usr/include/c++/13/bits/unique_ptr.h:242:7: note: candidate: ‘std::__uniq_ptr_data<_Tp, _Dp, <anonymous>, <anonymous> >::__uniq_ptr_data(std::__uniq_ptr_data<_Tp, _Dp, <anonymous>, <anonymous> >&&) [with _Tp = std::__future_base::_Result_base; _Dp = std::__future_base::_Result_base::_Deleter; bool <anonymous> = true; bool <anonymous> = true]’
  242 |       __uniq_ptr_data(__uniq_ptr_data&&) = default;
      |       ^~~~~~~~~~~~~~~
/usr/include/c++/13/bits/unique_ptr.h:242:7: note:   candidate expects 1 argument, 0 provided
make[4]: *** [CMakeFiles/mputil.dir/build.make:76: CMakeFiles/mputil.dir/src/mp/util.cpp.o] Error 1
```

</p>
</details>

--- a/depends/packages/capnp.mk
+++ b/depends/packages/capnp.mk
@@ -4,12 +4,17 @@ $(package)_download_path=$(native_$(package)_download_path)
 $(package)_download_file=$(native_$(package)_download_file)
 $(package)_file_name=$(native_$(package)_file_name)
 $(package)_sha256_hash=$(native_$(package)_sha256_hash)
+$(package)_patches=abi_placement_new.patch
 
 define $(package)_set_vars :=
   $(package)_config_opts := -DBUILD_TESTING=OFF
   $(package)_config_opts += -DWITH_OPENSSL=OFF
   $(package)_config_opts += -DWITH_ZLIB=OFF
   $(package)_cxxflags += -fdebug-prefix-map=$($(package)_extract_dir)=/usr -fmacro-prefix-map=$($(package)_extract_dir)=/usr
+endef
+
+define $(package)_preprocess_cmds
+  patch -p2 < $($(package)_patch_dir)/abi_placement_new.patch
 endef
 
 define $(package)_config_cmds

--- a/depends/patches/capnp/abi_placement_new.patch
+++ b/depends/patches/capnp/abi_placement_new.patch
@@ -1,0 +1,71 @@
+From 74560f26f75dda4257dce541ca362a1e763b2971 Mon Sep 17 00:00:00 2001
+From: Ryan Ofsky <ryan@ofsky.org>
+Date: Thu, 6 Feb 2025 08:39:05 -0500
+Subject: [PATCH 1/1] Avoid gcc/clang ABI incompatibility caused by
+ PlacementNew
+
+GCC and clang do not use same calling convention for passing empty struct
+parameters. There is more information about this in
+https://itanium-cxx-abi.github.io/cxx-abi/cxx-abi-dev/archives/2015-December/002869.html
+
+Unfortunately this can create an issue in capnproto if it is built without
+optimizations in GCC, and the resulting static libraries are used in a clang
+program, or vice versa.
+
+Depending on what order libraries are specified on the linker command line, and
+whether code compiled with the other compiler is calling any header functions
+that cause weak a `operator new(unsigned int, kj::_::PlacementNew, void*)`
+symbol to be defined in its own objects, this can cause the linker to link a
+GCC-generated `kj::ctor` with a clang-generated `operator new`, and the
+resulting program to crash due to the compilers using different calling
+conventions for `operator new`.
+
+This problem is difficult to avoid in general, but pretty easy to avoid here by
+changing `operator new` parameter order so the empty struct parameter is last.
+
+This change should be beneficial for capnproto users that may be compiling it
+without optimizations, and not necessarily using a single compiler to build all
+their dependencies.
+
+The problem does not occur if any optimizations are enabled because `operator
+new` calls are inlined in that case.
+---
+ c++/src/kj/common.h | 11 +++++++----
+ 1 file changed, 7 insertions(+), 4 deletions(-)
+
+diff --git a/c++/src/kj/common.h b/c++/src/kj/common.h
+index b8edde3c..28ab11d6 100644
+--- a/c++/src/kj/common.h
++++ b/c++/src/kj/common.h
+@@ -1034,24 +1034,27 @@ private:
+ 
+ // We want placement new, but we don't want to #include <new>.  operator new cannot be defined in
+ // a namespace, and defining it globally conflicts with the definition in <new>.  So we have to
+-// define a dummy type and an operator new that uses it.
++// define a dummy type and an operator new that uses it.  The dummy type is intentionally passed
++// as the last parameter so clang and GCC ABI calling conventions for empty struct struct parameters
++// are compatible, and there are not segfaults trying to call clang operator new/delete from GCC or
++// vice versa.
+ 
+ namespace _ {  // private
+ struct PlacementNew {};
+ }  // namespace _ (private)
+ } // namespace kj
+ 
+-inline void* operator new(size_t, kj::_::PlacementNew, void* __p) noexcept {
++inline void* operator new(size_t, void* __p, kj::_::PlacementNew) noexcept {
+   return __p;
+ }
+ 
+-inline void operator delete(void*, kj::_::PlacementNew, void* __p) noexcept {}
++inline void operator delete(void*, void* __p, kj::_::PlacementNew) noexcept {}
+ 
+ namespace kj {
+ 
+ template <typename T, typename... Params>
+ inline void ctor(T& location, Params&&... params) {
+-  new (_::PlacementNew(), &location) T(kj::fwd<Params>(params)...);
++  new (&location, _::PlacementNew()) T(kj::fwd<Params>(params)...);
+ }
+ 
+ template <typename T>

